frontends/ol-search-ui/src/components/TruncatedText.tsx

```diff
@@ -0,0 +1,108 @@
+import React, { useState, useRef, useLayoutEffect } from "react"
+
+import Dotdotdot from "react-dotdotdot"
+
+type Props = {
+  text: string
+  lines: number
+  // Estimated maximum number of characters on a single line of text for the element where this
+  // text will be rendered. Doesn't have to be exact.
+  estCharsPerLine: number
+  className?: string
+  showExpansionControls: boolean
+}
+
+interface DotDotDotElementWithContainer extends Dotdotdot {
+  container: MinimalDotDotDotContainerProps
+}
+
+type MinimalDotDotDotContainerProps = {
+  scrollHeight: number
+  offsetHeight: number
+}
+
+type RenderedNewlinesProps = {
+  text: string
+}
+
+const RenderedNewlines = React.memo<RenderedNewlinesProps>(({ text }) => (
+  <React.Fragment>
+    {text.split(/\n/g).map((item, key) => (
+      <React.Fragment key={key}>
+        {key === 0 ? null : <br />}
+        {item}
+      </React.Fragment>
+    ))}
+  </React.Fragment>
+))
+
+export default function TruncatedText(props: Props) {
+  const { text, lines, estCharsPerLine, className, showExpansionControls } =
+    props
+  const [expanded, setExpanded] = useState(false)
+  const [hasOverflow, setHasOverflow] = useState(false)
+  const [, forceReRender] = useState(0)
```

Comment:
This requires calling `forceReRender(val => val + 1)`.

A slightly nicer way to do it is
```
const [ignored, forceUpdate] = useReducer(x => x + 1, 0);
```
See https://reactjs.org/docs/hooks-faq.html#is-there-something-like-forceupdate